runtime/jcl/common/jclvm.c

```diff
@@ -138,6 +152,166 @@ Java_com_ibm_oti_vm_VM_allInstances(JNIEnv * env, jclass unused, jclass clazz, j
+{
+	J9VMThread *vmThread = (J9VMThread *) env;
+	J9JavaVM *vm = vmThread->javaVM;
+	J9InternalVMFunctions *vmFuncs = vm->internalVMFunctions;
+	J9HashTable *statsTable = NULL;
+	J9HashTableState hashTableState;
+	BOOLEAN outOfMemory = FALSE;
+	j9object_t stringObject = NULL;
+	J9HeapStatisticsTableEntry **statsArray = NULL;
+	jstring stringObjectRef = NULL;
+	I_32 numClasses = 0;
+
+	PORT_ACCESS_FROM_ENV(env);
+
+	vmFuncs->internalEnterVMFromJNI(vmThread);
+
+	vmFuncs->acquireExclusiveVMAccess(vmThread);
+	statsTable = collectHeapStatistics(vmThread);
+	vmFuncs->releaseExclusiveVMAccess(vmThread);
+
+	if (NULL != statsTable) {
+		numClasses = hashTableGetCount(statsTable);
+		statsArray = j9mem_allocate_memory(numClasses * sizeof(J9HeapStatisticsTableEntry*), J9MEM_CATEGORY_VM_JCL);
+	}
+	if (NULL == statsArray) {
+		outOfMemory = TRUE;
+	} else {
+		I_32 cursor = 0;
+		UDATA printedLength = 0;
+		UDATA bufferSize = 0;
+		J9HeapStatisticsTableEntry *entry = (J9HeapStatisticsTableEntry *) hashTableStartDo(statsTable, &hashTableState);
+		/* build a list of pointers to the hash table entries */
+		while (NULL != entry) {
+			entry->aggregateSize = entry->objectSize * entry->objectCount;
+			statsArray[cursor] = entry;
+			cursor += 1;
+			entry = (J9HeapStatisticsTableEntry *) hashTableNextDo(&hashTableState);
+		}
+		numClasses = cursor; /* adjust the length in case the hash table contained nulls */
+		qsort(statsArray, numClasses, sizeof(J9HeapStatisticsTableEntry*), compareByAggregateSize);
+		do {
+			char *stringBuffer = NULL;
+			bufferSize += numClasses * 80; /* try incrementally larger sizes */
+			stringBuffer = (char *) j9mem_allocate_memory(bufferSize, J9MEM_CATEGORY_VM_JCL);
+			if (NULL == stringBuffer) {
+				outOfMemory = TRUE;
+				break;
+			}
+			printedLength = printHeapStatistics(env, statsArray, numClasses, stringBuffer, bufferSize);
+			if (printedLength > 0) {
+				stringObject = vm->memoryManagerFunctions->j9gc_createJavaLangString(vmThread,
+						(U_8 *) stringBuffer, printedLength, J9_STR_XLAT);
+				stringObjectRef = vmFuncs->j9jni_createLocalRef(env, stringObject);
+			}
+			j9mem_free_memory(stringBuffer);
+		} while (0 == printedLength);
+		/* Need to keep the table until this point since statsArray contained pointer to its members. */
+		hashTableFree(statsTable);
+		j9mem_free_memory(statsArray);
+	}
+
+	if (outOfMemory) {
+		Trc_JCL_heapStatisticsOOM(vmThread);
+		vm->internalVMFunctions->setNativeOutOfMemoryError(vmThread, 0, 0);
+	}
+	vm->internalVMFunctions->internalExitVMToJNI(vmThread);
+
+	return stringObjectRef;
+}
+
+/**
+ * Arguments are pointers to entries in the statsArray,
+ * which are themselves pointers to entry in the hash table.
+ * Compare such that the list is sorted in descending order.
+ */
+static int
+compareByAggregateSize(const void *a, const void *b)
+{
+	const J9HeapStatisticsTableEntry *aEntry = *(const J9HeapStatisticsTableEntry* const *) a;
```

Comment:
Types don't match. Please use const on the proper side, particularly when using complex multi-indirect cases.

Reply:
The types *do* match.